components/org.wso2.carbon.crypto.provider/src/main/java/org/wso2/carbon/crypto/provider/SymmetricKeyInternalCryptoProvider.java

```diff
@@ -131,17 +153,111 @@ public SymmetricKeyInternalCryptoProvider(String secretKey) {
+            log.debug(String.format("Encrypting data with symmetric key encryption with algorithm: '%s'.", algorithm));
+        }
+        if (AES_GCM_SYMMETRIC_CRYPTO_ALGORITHM.equals(algorithm)) {
+            return encryptWithGCMMode(cleartext, javaSecurityAPIProvider);
+        }
+        return encrypt(cleartext, algorithm, javaSecurityAPIProvider);
+
+    }
+
+    private SecretKeySpec getSecretKey() {
+
+        return new SecretKeySpec(secretKey.getBytes(), 0, secretKey.getBytes().length,
+                DEFAULT_SYMMETRIC_CRYPTO_ALGORITHM);
+    }
+
+    private byte[] getInitializationVector() {
+
+        byte[] iv = new byte[GCM_IV_LENGTH];
+        UUID timeBasedUUID = Generators.timeBasedGenerator().generate();
+        ByteBuffer byteBuffer = ByteBuffer.wrap(iv);
+        byteBuffer.putLong(timeBasedUUID.getMostSignificantBits());
+        byteBuffer.putLong(timeBasedUUID.getLeastSignificantBits());
+        return byteBuffer.array();
+    }
+
+    /**
+     * This method will encrypt a given plain text in AES/GCM/NoPadding cipher transformation
+     *
+     * @param plaintext               plain text that need to be encrypted in this mode.
+     * @param javaSecurityAPIProvider crypto provider
+     * @return byte array of encrypted and self contained cipher text, which include cipher text and iv value.
+     * @throws CryptoException
+     */
+    private byte[] encryptWithGCMMode(byte[] plaintext, String javaSecurityAPIProvider)
+            throws CryptoException {
+
+        Cipher cipher;
+        byte[] cipherText;
+        byte[] iv = getInitializationVector();
+        try {
+            if (StringUtils.isBlank(javaSecurityAPIProvider)) {
+                cipher = Cipher.getInstance(AES_GCM_SYMMETRIC_CRYPTO_ALGORITHM);
+            } else {
+                cipher = Cipher.getInstance(AES_GCM_SYMMETRIC_CRYPTO_ALGORITHM, javaSecurityAPIProvider);
+            }
+            cipher.init(Cipher.ENCRYPT_MODE, getSecretKey(), getGCMParameterSpec(iv));
+            cipherText = cipher.doFinal(plaintext);
+            cipherText = createSelfContainedCiphertext(cipherText, AES_GCM_SYMMETRIC_CRYPTO_ALGORITHM, iv);
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | InvalidAlgorithmParameterException | IllegalBlockSizeException | BadPaddingException | NoSuchProviderException e) {
+
+            String errorMessage = String.format("Error occurred while initializing and encrypting using Cipher object" +
+                    " with algorithm: '%s'.", AES_GCM_SYMMETRIC_CRYPTO_ALGORITHM);
+            throw new CryptoException(errorMessage, e);
+        }
+        return cipherText;
+    }
+
+    /**
+     * This method is to create GCMParameterSpec which is needed to encrypt and decrypt operation with AES-GCM
+     * @param iv
+     * @return
+     */
+    private GCMParameterSpec getGCMParameterSpec(byte[] iv) {
+
+        return new GCMParameterSpec(GCM_TAG_LENGTH * 8, iv);
```

Comment:
Shall we move 8 to a constant as well and comment on the rationale behind that?

Reply:
Fixed with d6dc3257ddc2151fae53aa5d933c97f1c5942689